Objetivo 2

marcosrmartin opened the pull request:
Sobre la estructura del repositorio
* [x] ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
   disposición de los mismos?
Sobre el análisis del problema
* [ ] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado? En los commits.
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.
Sobre la planificación y la programación
* [x] ¿Los issues representan un problema, y no una tarea?
* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están asignados a una HU?
* [ ] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit? Si. ¿Han sido asignados al primer milestone? No, ya que los hemos rehecho.
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Son los milestones sobre los que estoy trabajando productos
   mínimamente viables? ¿O tengo que solicitar al product manager que
   cree milestones adicionales o precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?


Hola Daniel, estas son las estructuras de datos que he pensado que pueden resolver el problema con algunas comprobaciones de errores, si ves que se pueden añadir más lo hablamos.
Aparte de esto, la entidad Ruta he decidido que almacene una puntuación para la heurística, un tramo nuevo y un puntero a otro objeto Ruta, ya que así la estructura se podría usar recursivamente (para algoritmos como Dijkstra, y otros de VRP).
Si ves que hay algo mal dímelo, y le doy otra vuelta, en el caso de ser sobre la ED Ruta, abrimos un issue para discutirlo allí, de lo contrario espero la aprobación.🤝

marcosrmartin committed:
Archivo con el módulo mínimo.

marcosrmartin committed:
Declaración de las EDs de objetos-valor e identidad.

marcosrmartin committed:
Estructura corregida y main completado para compilación.

marcosrmartin committed:
Main update.

marcosrmartin committed:
iv.yaml update.

marcosrmartin committed:
Ajust nombre para coincidir con #3.

marcosrmartin committed:
Ajusto nombre heurística para coincidir con dmonjasm#3.

marcosrmartin committed:
close #8 close #9 close #10 debido a que están mal planteados desde el inicio.

marcosrmartin committed:
Lenguaje de programación close #7

marcosrmartin committed:
Implementación ED Objeto-valor Posicion #16.

marcosrmartin committed:
Implementación ED Objeto-valor Peaje # 14.

marcosrmartin committed:
Implementación ED Objeto-valor Vehiculo #15.

marcosrmartin committed:
Implementación ED Objeto-valor Tramo #13.

marcosrmartin committed:
Implementación ED Entidad Ruta #12.

marcosrmartin committed:
Fixed iv.yml #7

marcosrmartin committed:
Fix iv.yml #7

dmonjasm commented:
He estado revisando el código y está todo bien, menos en la implementación de Ruta, que creo que el atributo tramo debería ser una slice o un list ya que entiendo que una ruta tendrá varios tramos y no solo uno. Por otro lado falta añadir la inicialización del modulo (el go.mod) en el directorio raíz. Cambiado eso yo lo veo ya todo bien.

marcosrmartin committed:
Module PMV. #12

marcosrmartin commented:
Lo que he hecho es que una Ruta pueda tener otra Ruta dentro, y como cada Ruta tiene un tramo, ese seria la "list" de tramos, ya que se podría recorrer recursivamente.

dmonjasm commented:
vale ya veo lo que dices, pues entonces lo veo todo bien. Le doy el ok

dmonjasm reviewed (approved):
:+1:

marcosrmartin commented:
@JJ Listo para revisión.

Carlos-SE commented:
@marcosrmartin, creo que deberías cambiar los `issues` o crear unos nuevos ya que los que has planteado han sido titulados como si fuesen historias de usuario. También añadir que en los issues no tienes que proponer la solución, tienes que plantear el problema y avanzar sobre ellos con los `commits` adecuados.

JJ commented:
> He estado revisando el código y está todo bien, menos en la implementación de Ruta, que creo que el atributo tramo debería ser una slice o un list ya que entiendo que una ruta tendrá varios tramos y no solo uno. Por otro lado falta añadir la inicialización del modulo (el go.mod) en el directorio raíz. Cambiado eso yo lo veo ya todo bien.

Esto debes hacerlo inline en el código correspondiente. Por favor, aprended a hacer estas cosas.

JJ commented:
> @marcosrmartin, creo que deberías cambiar los `issues` o crear unos nuevos ya que los que has planteado han sido titulados como si fuesen historias de usuario. También añadir que en los issues no tienes que proponer la solución, tienes que plantear el problema y avanzar sobre ellos con los `commits` adecuados.

Esencialmente es esto mismo. Hay issues que están marcados como historias de usuario pero no lo son, no están relacionados con ninguna historia de usuario, y no plantean problemas. El que (más o menos) plantea uno, el #12 (que tampoco es una historia de usuario) sólo tiene un puñado de commits.

Por favor, seguid también las prácticas habituales en commits "Avanza #xx" no es lo adecuado.

JJ commented on internal/Posicion.go:
```diff
@@ -0,0 +1,22 @@
+package internal
+import "errors"
+
+type Posicion struct{
```
No parece que esto tenga ninguna relación con la lógica de negocio.

JJ commented on internal/Posicion.go:
```diff
@@ -0,0 +1,22 @@
+    return &Posicion{x, y}, nil
+}
+
+func PosicionError(x float32, y float32) error{
```
Y esto tampoco.

JJ commented on internal/Tramo.go:
```diff
@@ -0,0 +1,25 @@
+import "errors"
+
+type Tramo struct{
+    inicio Posicion
```
No creo que se vayan a buscar los tramos por coordenadas geográficas. No pone nada parecido en ninguna HU, ni parece que eso solucione el problema de representar los tramos para usarlos por la lógica de negocio.

JJ commented on internal/Peaje.go:
```diff
@@ -0,0 +1,29 @@
+import "errors"
+
+type Peaje struct{
+    precios map[Peso]float32
```
Más que peso debería ser tipo de vehículo, no creo que haya solo dos tipos.

JJ commented:
> Sobre la estructura del repositorio
> 
>     * [x]  ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
>       disposición de los mismos?

No estoy seguro de que `internal` sea lo más habitual. He visto `pkg`, por ejemplo, pero no esto.

>       Sobre el análisis del problema
> 
>     * [x]  ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
>       que se ha creado?
> 

Si no lo has hecho, no lo marques.

>     * [x]  ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
>       una entidad o un agregado?
> 

Tampoco.

JJ reviewed (changes requested):
Hay que seguir la metodología propuesta
* Asignar issues a milestones (por parte de @dmonjasm )
* No llamar issues a los que no lo son.
* Issues relacionados siempre con HUs
* Issues que planteen problemas.

Sin una base sólida de issues para trabajar, hay que rehacer gran parte del trabajo, porque el código desarrollado no se ha hecho correctamente.

dmonjasm commented:
> > Sobre la estructura del repositorio
> > ```
> > * [x]  ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
> >   disposición de los mismos?
> > ```
> 
> No estoy seguro de que `internal` sea lo más habitual. He visto `pkg`, por ejemplo, pero no esto.

Buscando las buenas prácticas de GO vi que un proyecto GO básico suele tener tres directorios cmd, pkg e internal. En este caso solo se ha creado internal porque es el directorio que contiene el código privado del proyecto. Adjunto enlace [ProjectStruture](https://dev.to/jinxankit/go-project-structure-and-guidelines-4ccm)

JJ commented:
> > > Sobre la estructura del repositorio
> > > ```
> > > * [x]  ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
> > >   disposición de los mismos?
> > > ```
> > 
> > 
> > No estoy seguro de que `internal` sea lo más habitual. He visto `pkg`, por ejemplo, pero no esto.
> 
> Buscando las buenas prácticas de GO vi que un proyecto GO básico suele tener tres directorios cmd, pkg e internal. En este caso solo se ha creado internal porque es el directorio que contiene el código privado del proyecto. Adjunto enlace [ProjectStruture](https://dev.to/jinxankit/go-project-structure-and-guidelines-4ccm)

¿Por qué es privado y no externo? ¿Cómo has modelado el código para decidir eso?

dmonjasm commented:
> > > > Sobre la estructura del repositorio
> > > > ```
> > > > * [x]  ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
> > > >   disposición de los mismos?
> > > > ```
> > > 
> > > 
> > > No estoy seguro de que `internal` sea lo más habitual. He visto `pkg`, por ejemplo, pero no esto.
> > 
> > 
> > Buscando las buenas prácticas de GO vi que un proyecto GO básico suele tener tres directorios cmd, pkg e internal. En este caso solo se ha creado internal porque es el directorio que contiene el código privado del proyecto. Adjunto enlace [ProjectStruture](https://dev.to/jinxankit/go-project-structure-and-guidelines-4ccm)
> 
> ¿Por qué es privado y no externo? ¿Cómo has modelado el código para decidir eso?

Entiendo que el modelado de las entidades y objetos valor de este proyecto son útiles sólo para este proyecto, luego por eso se mete en internal. Tenía entendido que en pkg se mete aquel código que puede ser útil para otro proyecto o servicios.

marcosrmartin committed:
Borramos todo para empezar de nuevo el objetivo-2 close #10 close #11 close #12 close #13 close #14 close #15 close #16

Carlos-SE commented:
Buenas, @marcosrmartin. Los issues ya tienen mejor pinta, sigue así y mucho ánimo! 👍

marcosrmartin committed:
Selección del lenguaje de programación close #7

marcosrmartin committed:
Implementación de EDs Ruta y Tramo close #18
Donde tramo será una estructura auxiliar para poder calcular el precio
de una ruta, la cual será una entidad.

marcosrmartin committed:
Implementación ED ConsumoVehiculo close #19
La cual será un objeto-valor, que junto con una Ruta sirve para calcular su precio.

JJ commented on internal/ruta.go:
```diff
@@ -0,0 +1,16 @@
+type ConsumoVehiculo float32
+
+type Ruta struct{
+	precio float32
```
¿Precio o coste?

JJ commented on internal/ruta.go:
```diff
@@ -0,0 +1,16 @@
+	if len(tramos) == 0 {
+		return nil, errors.New("Una ruta debe contener mínimo un tramo.")
+	}
+	return &Ruta{0, tramos}, nil
```
No sé si el PMV incluye inicializar el coste, pero si lo hace, ¿por qué es cero siempre?
Si modelizar tramos de autopistano es de este PMV, ¿por qué lo pones?

marcosrmartin replied:
No lo tiene que inicializar, pero en go los enteros no se pueden inicializar a nulo, su equivalente es 0, o se podría utilizar un valor negativo. En cuanto a los tramos los propongo porque no puedes calcular una ruta sin saber de que está compuesta.

JJ replied:
¿Y por qué no lo pasa al constructor entonces?
En todo caso, ¿por qué no se ha planteado eso en un issue?

marcosrmartin replied:
Miento, el coste se debería de calcular con otra función, pero esta debería de estar incluida en el constructor, aunque llevaría a cabo la lógica de negocio, pero como no es está incluido en este PMV lo inicializo a ""nulo"". En cuanto a porque no se ha planteado tramos como un issue, es porque lo entiendo como una solución al problema de representar una Ruta, y en los issues no se deben proponer soluciones, realmente uso tramo porque en Go no existen tuplas y uso esa clase como apoyo para Ruta.

JJ reviewed (changes requested):
Los issues no plantean realmente ningún problema, y no se ha hecho ningún intento por modelizar el mismo. Por ejemplo, no se da ninguna pista sobre como se va a llevar a cabo #3 o como esto contribuye a eso.

marcosrmartin committed:
Implementación de Objeto-valor Tramo #20.
Cambio nombre límite a velocidadLimite para cumplir DDD.

marcosrmartin committed:
Añado peaje, y elimino el coste  #18
Para que ruta sea un Objeto-valor y se calcule desde fuera.

marcosrmartin commented:
@JJ Listo para revisión.

JJ reviewed (approved):
No has justificado mucho cómo has hecho el modelo, pero lo hemos hablado en clase, así que por mi :+1:
